Mvc pattern

vaberof opened the pull request.

vaberof committed:
refactor

vaberof committed:
refactor MVC

goodmove commented on controller/bot.go:
```diff
@@ -0,0 +1,69 @@
+	"log"
+)
+
+const TOKEN = ""
```
это лучше в конфиг какой-нибудь унести (.properties, .ini, .yaml). Можно посмотреть [либу](https://pkg.go.dev/go.uber.org/config)

goodmove commented on controller/bot.go:
```diff
@@ -0,0 +1,69 @@
+func HandleMessage(userText string) *string {
+	var response string
+
+	groups := model.CreateGroupStorage()
```
почему Storage создается заново на каждое сообщение пользователя?

goodmove commented on controller/bot.go:
```diff
@@ -0,0 +1,69 @@
+		return &response
+	}
+
+	schedule := model.GetTodaySchedule(studyGroupUrl)
```
вот здесь более корректно будет получить сначала доменный schedule, а потом сериализовать его в нужный вид

goodmove commented on controller/bot.go:
```diff
@@ -0,0 +1,69 @@
+
+		inputText := update.Message.Text
+		responseText := HandleMessage(inputText)
+		responseMessage.Text = *responseText
```
а если `responseText == nil`?

goodmove commented on handler/http.go:
```diff
@@ -0,0 +1,37 @@
+	if res.StatusCode != 200 {
+		fmt.Println("status code error: %d %s", res.StatusCode, res.Status)
+	}
+	body := res
```
эта переменная не очень понятно зачем нужна

goodmove commented on handler/http.go:
```diff
@@ -0,0 +1,37 @@
+		log.Println(err)
+	}
+
+	//defer res.Body.Close()
```
это можно убрать

goodmove commented on handler/http.go:
```diff
@@ -0,0 +1,37 @@
+	"net/http"
+)
+
+// Делаем запрос
```
комментарии достаточно бесполезные)

goodmove commented on controller/bot.go:
```diff
@@ -0,0 +1,69 @@
+	return bot
+}
+
+// Получаем сообщение от пользователя
```
это же не так. Комментарий неверный

goodmove commented on model/date.go:
```diff
@@ -0,0 +1,33 @@
+	var shortTodayDate string
+	var fullTodayDate string
+
+	shortTodayDate = time.Now().In(setLocation()).Format("02.01")
```
достаточно узнать время один раз, а затем конвертировать его в несколько форматов

goodmove replied:
location нужно где-то вычислить один раз и запомнить: нет смысла при каждом получении дат создавать новый location

goodmove commented on model/date.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+// Получаем сегодняшнюю дату
+func GetTodayDate() []string {
```
тут лучше структуру вернуть, потому что массив не помогает понять, что за данные там лежат, как обрабатывать 0 элементов, 1 или больше 2

goodmove commented on model/group.go:
```diff
@@ -1,39 +1,48 @@
-package models
+package model
 
 // http://rasp.sgugit.ru/?ii=1&fi=1&c=1&gn=1044& БИ-11.1
```
а что за комменты?

goodmove commented on model/group.go:
```diff
@@ -1,39 +1,48 @@
 
+// Получаем url адрес группы
+func (g *GroupStorage) GetGroupUrl(group string) string {
+	url, _ := g.groups[group]
```
а если такой группы нет?

goodmove commented on model/group.go:
```diff
@@ -1,39 +1,48 @@
 }
 
+// Получаем url адрес группы
+func (g *GroupStorage) GetGroupUrl(group string) string {
```
что такое `group`?

goodmove commented on model/group.go:
```diff
@@ -1,39 +1,48 @@
-func (g *Groups) GetGroup(group string) string {
-	url, _ := g.groups[group]
+// Проверяем, существует ли группа, введенная пользователем
+func (g *GroupStorage) isSaved(group string) bool {
```
нигде не используеся

goodmove commented on model/group.go:
```diff
@@ -1,39 +1,48 @@
 // Проверяем, существует ли группа, введенная пользователем
-func (g *Groups) GroupExists(group string) bool {
+func (g *GroupStorage) Exists(group string) bool {
 	if _, exist := g.groups[group]; exist {
```
```go
_, exists := g.groups[group]

return exists

```

goodmove commented on model/schedule.go:
```diff
@@ -0,0 +1,74 @@
+package model
+
+type Schedule struct {
```
schedule должен описывать расписание по дням. Что сейчас такое `[]string`? как из этого получить расписание на конкретный день? Как работать с этим? Почему нельзя просто `[]string` использовать?

goodmove commented on model/schedule.go:
```diff
@@ -0,0 +1,74 @@
+	schedule []string
+}
+
+func CreateSchedule() *Schedule {
```
в go соглашение использовать `NewSchedule` или просто `New` для таких операций

goodmove commented on model/schedule.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+// Добавляем пары в массив
+func (s *Schedule) AddLessons(startTime, finishTime, lessonName, roomNumber, teacherName, lessonType string) {
```
нужно разделить сериализацию и домен

goodmove commented on model/schedule.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+// Возвращаем расписание пользователю
+func (s *Schedule) GetSchedule() string {
```
функция называется так, будто возвращается константа. Лучше явно назвать `GetScheduleForToday` или дату добавить параметром

goodmove replied:
ну и у тебя есть тип Schedule, а тут почему-то возвращается string

goodmove commented on model/schedule.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+// Добавляем дату в начало массива
+func (s *Schedule) AddTodayDate() {
```
`GetTodayDate()[1]` – вот это точно сломается в один прекрасный момент. Если в списке будет 0 элементов, что делать?

goodmove commented on model/schedule.go:
```diff
@@ -0,0 +1,74 @@
+
+// Проверяем, есть ли сегодня пары
+func (s *Schedule) todayScheduleExists(schedule []string) bool {
+	if len(schedule) == 0 {
```
`return len(schedule) != 0`

goodmove commented on model/parser.go:
```diff
@@ -0,0 +1,93 @@
+
+	// Проверяем, существует ли тег на странице
+	if isNilSelection(dateSelection) {
+		responseMessage = "Воскресенье - пар нет"
```
есть какая-то гарантия, что в этот блок можно попасть только в воскресенье?

goodmove commented on model/parser.go:
```diff
@@ -0,0 +1,93 @@
+		responseMessage string // Сообщение пользователю
+	)
+
+	schedule := CreateSchedule()
```
объект создается впустую, если сегодня воскресенье

vaberof committed:
refactor

vaberof committed:
fix

vaberof committed:
fix

1. add .idea to .gitignore
2. rewrite comments

vaberof committed:
fix

1. change path to .yaml file
2. rename function 'GroupUrl' to 'StudyGroup'
3. rewrite comments

vaberof committed:
update

1. add new func 'dayTranslate'
2. change style of returnable schedule to user

vaberof committed:
bug fix

fixed bug when user input group id from current chat
were changing by user from another chat with his input group id,
because of which user were getting wrong schedule.

vaberof committed:
update

1. change name of module in go.mod
2. rename func 'AddUser' to 'AddData'
3. rewrite comments

vaberof committed:
update

1. Added function to parse every date of the week days
2. Added function to parse lessons for the week
3. Added function to get dates of the week days
4. Added function to call correct parser
5. Changed project-layout

goodmove commented on internal/app/bot/client.go:
```diff
@@ -0,0 +1,65 @@
+// Start starts the bot.
+func Start() {
+
+	var (
```
зачем скобки тут? это очень тяжело читать. тут одна переменная и это не нужно (и лучше вообще нигде не нужно)


Называй все переменные максимально точно
```go
botKeyboardMarkup := tgbotapi.NewInlineKeyboardMarkup(...)
```

goodmove commented on internal/app/bot/client.go:
```diff
@@ -0,0 +1,65 @@
+	var (
+		inlineKeyboard = tgbotapi.NewInlineKeyboardMarkup(
+			tgbotapi.NewInlineKeyboardRow(
+				tgbotapi.NewInlineKeyboardButtonData("Сегодня", "Сегодня"),
```
почему здесь два русский текста? зачем этот текст?

goodmove commented on internal/app/bot/client.go:
```diff
@@ -0,0 +1,65 @@
+
+	bot := NewBot()
+
+	channel := tgbotapi.NewUpdate(0)
```
что за `channel`? почему здесь 0 передается в аргументах?

goodmove commented on internal/app/bot/client.go:
```diff
@@ -0,0 +1,65 @@
+	channel := tgbotapi.NewUpdate(0)
+	channel.Timeout = 60
+
+	updates := bot.GetUpdatesChan(channel)
```
`botUpdatesChannel`

goodmove commented on internal/app/bot/client.go:
```diff
@@ -0,0 +1,65 @@
+
+	updates := bot.GetUpdatesChan(channel)
+
+	user := model.NewUser()
```
откуда здесь `user`? еще ни одного сообщения от пользователей не пришло

goodmove commented on internal/app/bot/client.go:
```diff
@@ -0,0 +1,65 @@
+
+	user := model.NewUser()
+	groupStorage := model.NewGroupStorage()
+	location := model.SetLocation()
```
лучше назвать `GetDefaultLocation`. Никакой location здесь нигде не устанавливается

goodmove commented on internal/app/bot/client.go:
```diff
@@ -0,0 +1,65 @@
+
+	for update := range updates {
+
+		if update.CallbackQuery != nil {
```
все лучше организовать более просто:

```go
for update := range updates {
  if (menuButtonPressed) {
    handleMenuButtonPress(...)
  } else if (messageReceived) {
    handleNewMessage(...)
  }
}

```

goodmove commented on internal/app/bot/client.go:
```diff
@@ -0,0 +1,65 @@
+		inputText := responseMessage.Text
+		chatID := responseMessage.ChatID
+
+		user.AddData(chatID, inputText)
```
AddData слишком общее название, лучше как везде что-то конкретное писать

goodmove commented on internal/app/bot/client.go:
```diff
@@ -0,0 +1,65 @@
+	groupStorage := model.NewGroupStorage()
+	location := model.SetLocation()
+
+	for update := range updates {
```
в общем не понятно, чем отличаются ветки алгоритма.  везде дублируется код с какими-то деталями и хз в чем разница обработки данных

goodmove commented on internal/app/controller/scheduleController.go:
```diff
@@ -0,0 +1,29 @@
+// returns schedule for user.
+// If user`s input group not exists,
+// returns a corresponding message.
+func HandleMessage(userChatID int64, date string, user *model.User, studyGroupStorage *model.GroupStorage, location *time.Location) *string {
```
так почему здесь возвращается не Schedule, а какой-то указатель на строку?

goodmove replied:
дата должна быть датой, а не строкой

goodmove replied:
location тут не нужен

goodmove commented on internal/app/controller/scheduleController.go:
```diff
@@ -0,0 +1,29 @@
+func HandleMessage(userChatID int64, date string, user *model.User, studyGroupStorage *model.GroupStorage, location *time.Location) *string {
+	var response string
+
+	studyGroupId := user.Data[userChatID]
```
1. опять не понятно, что такое `Data`. 
2. не понятно, откуда тут появляются какие-то  данные, учитывая, что handleMessage в основном цикле вызывается до обновления этой мапки
3. это не должно называться `user`. Это же не информация о каком-то одном пользователе, а фактически БД с какими-то данными всех пользователей

goodmove commented on internal/app/controller/scheduleController.go:
```diff
@@ -0,0 +1,29 @@
+	}
+
+	schedule := services.GetSchedule(studyGroupId, date, *studyGroupUrl, location)
+	scheduleString := services.ScheduleToString(&schedule)
```
это нужно вынести в тело основного цикла

goodmove commented on internal/app/controller/scheduleController.go:
```diff
@@ -0,0 +1,29 @@
+		return &response
+	}
+
+	schedule := services.GetSchedule(studyGroupId, date, *studyGroupUrl, location)
```
почему тут не подходит передать указатель `studyGroupId`? зачем разыменование?

goodmove commented on internal/app/handler/http.go:
```diff
@@ -0,0 +1,42 @@
+}
+
+// LoadHtmlPage loads html page.
+func LoadHtmlPage(url string) *goquery.Document{
```
публичные функции лучше всегда располагать вверху, чтобы быстрее понимать, что экспортируется и не листать весь файл в поисках функций, которые можно использовать у себя

goodmove commented on internal/app/model/date.go:
```diff
@@ -0,0 +1,111 @@
+
+const location = "Asia/Novosibirsk"
+
+type Date struct {
```
очень странный тип с абсолютно непонятной семантикой. Можешь назвать две недели в году, когда имена дней отличаются?

goodmove replied:
и не понятно, почему у даты есть дни недели

goodmove replied:
есть просто дата: число, день недели, имя дня недели и все. Есть тип `Time`. Лучше его использовать везде и все, а сериализацию отдельно проводить

goodmove commented on internal/app/model/date.go:
```diff
@@ -0,0 +1,111 @@
+	todayDate := time.Now().In(location)
+	day := todayDate.Weekday()
+
+	d.shortDate = todayDate.Format("02.01")
```
зачем алгоритм форматирования дублировать везде?

goodmove commented on internal/app/model/date.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+// dayTranslate translates day of the week from english to russian.
+func dayTranslate(day string) string {
```
это нужно назвать `func WeekDayEngRu(engDay string) string`

goodmove commented on internal/app/services/scheduleService.go:
```diff
@@ -0,0 +1,40 @@
+// dateSchedule calls ParseDayLessons or ParseWeekLessons depending on user input
+func dateSchedule(groupId, date, url string, location *time.Location) model.Schedule {
+	switch date {
+	case "Сегодня":
```
все кейсы нужно вынести в константы и использовать глобально

goodmove commented on internal/app/services/scheduleService.go:
```diff
@@ -0,0 +1,40 @@
+	switch date {
+	case "Сегодня":
+		return *model.ParseDayLessons(groupId, date, url, location)
+	case "Завтра":
```
```go
case A, B, C, D:
  doX()
default:
  doY()

```

goodmove commented on internal/app/services/scheduleService.go:
```diff
@@ -0,0 +1,40 @@
+	case "След. неделя":
+		return *model.ParseWeekLessons(groupId, date, url, location)
+	default:
+		return model.Schedule{}
```
какая у этого семантика?

goodmove commented on internal/app/services/scheduleService.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+// dateSchedule calls ParseDayLessons or ParseWeekLessons depending on user input
+func dateSchedule(groupId, date, url string, location *time.Location) model.Schedule {
```
нужно возвращать указатель, чтобы не копировать структуру

goodmove commented on internal/app/model/schedule.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// AddLessons adds lessons to schedule.
+func (s *Schedule) AddLessons(startTime, finishTime, lessonName, roomNumber, teacherName, lessonType string) {
```
это все нужно унести в сериализацию. это не относится к модели

goodmove commented on internal/app/model/schedule.go:
```diff
@@ -0,0 +1,78 @@
+package model
+
+type Schedule struct {
```
поля в структуре не имеют никакого смысла

goodmove commented on internal/app/model/parser.go:
```diff
@@ -0,0 +1,173 @@
+package model
```
я бы структурировал приложение так:

```
cmd/app/main.go

internal/bot/
- config.go
- bot..go

internal/
- pkg/
  - http/http.go # http utils
  - date/date.go # date utils
- integration/unisite/ # http integration with university website
  - api.go # api definition to get schedule
  - parser.go # html parser
- app/
  - handler/schedule.go # bot message handler
  - model/
    - schedule.go # schedule classes definition
    - date.go # may be redundant
  - service/scheduleService.go
  - storage/groupStorage.go

```

goodmove replied:
модель расписания лучше представить так:

```go


type Schedule = map[Date]DaySchedule
type DaySchedule = []Lesson
type Lesson struct {
  Name string
  TeacherFullName string
  Room string
}


```

vaberof committed:
refactor

vaberof committed:
create command handler

vaberof committed:
refactor

vaberof committed:
add Procfile

vaberof committed:
change procfile

vaberof committed:
fix procfile

vaberof committed:
fix procfile

vaberof committed:
fix procfile

vaberof committed:
fix procfile

vaberof committed:
set webhook

vaberof committed:
set environment variable

vaberof committed:
set a port

vaberof committed:
fix webhook

vaberof committed:
create procfile

vaberof committed:
fix procfile

vaberof committed:
fix procfile

vaberof committed:
fix procfile

vaberof committed:
delete procfile

vaberof committed:
set up procfile

vaberof committed:
fix procfile

vaberof committed:
fix procfile

vaberof committed:
fix procfile

vaberof committed:
fix procfile

vaberof committed:
remove procfile

vaberof committed:
create procfile

vaberof committed:
fix procfile

vaberof committed:
fix

vaberof committed:
delete loadConfig func

vaberof committed:
test procfile

vaberof committed:
create executable file

vaberof committed:
test procfile

vaberof committed:
fix procfile

vaberof committed:
test procfile

vaberof committed:
test procfile

vaberof committed:
refactor

added http timeout and http error handler

vaberof committed:
implemented a cache

vaberof committed:
rewrite names of variables

vaberof committed:
changed the struct of logs

vaberof committed:
refactor

vaberof committed:
fix schedule storage expire time

create func SetNewExpireTime
create func ClearSchedule

vaberof committed:
refactor

1. add custom http error which caused by timeout
2. rewrite comments

vaberof committed:
add README.md file

vaberof committed:
Update README.md

vaberof committed:
Update README.md

goodmove commented on internal/app/handler/message.go:
```diff
@@ -0,0 +1,21 @@
+// HandleMessage
+// returns group id, url to parse schedule and error.
+// if user`s input group id not exists, then error value becomes custom error.
+func HandleMessage(chatID int64, msgStorage *storage.MessageStorage, groupStorage *storage.GroupStorage) (string, string, error) {
```
а зачем нужна эта обертка? почему там где нужно не достать из хранилища нужную группу?

goodmove commented on internal/app/model/schedule.go:
```diff
@@ -0,0 +1,49 @@
+package model
+
+type Schedule map[string]DaySchedule
```
здесь для ключа я б тоже сделал какой-нибудь тип, а-то не понятно что такое `string`

goodmove replied:
и я б наверное для простоты сделал `type Schedule = map[Something]DaySchedule`

goodmove commented on internal/app/model/schedule.go:
```diff
@@ -0,0 +1,49 @@
+package model
+
+type Schedule map[string]DaySchedule
+type DaySchedule []Lesson
```
тут лучше сделать `[]*Lesson`, чтобы не нужно было большой кусок памяти выделять и копировать данные

goodmove commented on internal/app/model/schedule.go:
```diff
@@ -0,0 +1,49 @@
+	return &DaySchedule{}
+}
+
+func (d *DaySchedule) AddLessons(lessonName, startTime, finishTime, lessonType, roomNumber, teacherName string) {
```
почему AddLesson**s**, когда добавляется только один?

goodmove replied:
кажется, что это вообще не нужно. Нужно сделать `ScheduleStorage` и там сделать метод `SaveLesson(scheduleId, weekday, lesson)`

goodmove commented on internal/app/service/scheduleService.go:
```diff
@@ -0,0 +1,289 @@
+	daySchedule := dereferenceSchedule[inputCallback]
+
+	for i := 0; i < len(daySchedule); i++ {
+		lessonData := reflect.ValueOf(daySchedule[i])
```
зачем тут reflection?

goodmove commented on internal/app/service/scheduleService.go:
```diff
@@ -0,0 +1,289 @@
+package service
```
в `go` все файлы принято именовать в `snake_case`, но в целом не критично

goodmove replied:
и сервиса-то в итоге никакого нет

goodmove commented on internal/app/model/schedule.go:
```diff
@@ -0,0 +1,49 @@
+	TeacherFullName string
+}
+
+func NewDaySchedule() *DaySchedule {
```
мне кажется, это бессмысленная функция. Ничего не делает

goodmove commented on internal/pkg/date/date.go:
```diff
@@ -0,0 +1,70 @@
+	"github.com/vaberof/goweekdate"
+)
+
+func GetParseDate(inputCallBack string) time.Time {
```
в названии функции ошибка по-моему какая-то. Get или Parse в итоге?

goodmove replied:
что такое `input callback`?

goodmove replied:
есть `time.ParseDate`. Почему этого не достаточно?

goodmove commented on internal/pkg/date/date.go:
```diff
@@ -0,0 +1,70 @@
+	log "github.com/sirupsen/logrus"
+	"time"
+
+	"github.com/vaberof/TelegramBotUniversitySchedule/internal/constants"
```
не должен внешний пакет `pkg` зависеть от внутреннего пакета `internal`. Это просто идеологически криво)

goodmove commented on internal/pkg/utils/WeekDayEngRu.go:
```diff
@@ -0,0 +1,27 @@
+
+// WeekDayEngRu translates day of the week from english to russian.
+func WeekDayEngRu(engDay time.Weekday) string {
+	switch engDay.String() {
```
зачем здесь эти константы, когда weekday – это и есть константа

```go
func FormatRuWeekday(weekday time.Weekday) string {
  switch weekday {
    case time.Monday: 
      return "Понедельник"
  }
}

```

goodmove replied:
и старайся никогда не называть пакеты utils – это сразу превращается в помойку с ненужным кодом

goodmove commented on internal/pkg/http/http.go:
```diff
@@ -0,0 +1,77 @@
+
+// makeRequest returns response body of type Reader
+// and error if http request Timeout occurred or response body cannot be read.
+func makeRequest(url string) (io.Reader, error) {
```
рекомендую взять вот эту либу: https://github.com/go-resty/resty

там уже куча инструментов готовых и все красиво работает

goodmove commented on internal/pkg/date/date.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+func GetCurrentTime() time.Time {
+	return time.Now().In(GetDefaultLocation(constants.Location))
```
тут же panic будет, когда `GetDefaultLocation` вернет nil

goodmove commented on internal/pkg/date/date.go:
```diff
@@ -0,0 +1,70 @@
+
+func GetDefaultLocation(location string) *time.Location {
+	loc, err := time.LoadLocation(location)
+	if err != nil {
```
почему err не возвращается?

goodmove commented on internal/pkg/date/date.go:
```diff
@@ -0,0 +1,70 @@
+	return Dates
+}
+
+func GetDefaultLocation(location string) *time.Location {
```
сигнатура должна быть `(*time.Location, error)`

goodmove commented on internal/integration/unisite/parser.go:
```diff
@@ -0,0 +1,194 @@
+	"strings"
+	"time"
+
+	"github.com/vaberof/TelegramBotUniversitySchedule/internal/app/model"
```
integration не должен зависеть от доменных классов

goodmove commented on cmd/app/main.go:
```diff
@@ -0,0 +1,9 @@
+)
+
+func main() {
+	bot.Start()
```
Я бы добавил все-таки конфиг приложения. Можно взять https://github.com/spf13/viper

goodmove commented on go.mod:
```diff
@@ -1,10 +1,20 @@
 
 go 1.17
 
 require (
```
кажется, тебе нужно сделать `go mod tidy`

goodmove commented on go.mod:
```diff
@@ -1,10 +1,20 @@
-module github.com/tg_bot_timetable
+module github.com/vaberof/TelegramBotUniversitySchedule
 
 go 1.17
```
рекомендую обновиться до 1.18

goodmove commented on internal/bot/bot.go:
```diff
@@ -0,0 +1,204 @@
+)
+
+// Start starts the bot.
+func Start() {
```
тело этой функции можно полностью унести в main.go

goodmove commented on internal/bot/bot.go:
```diff
@@ -0,0 +1,204 @@
+// handleNewMessage
+// adds user`s chat id and his input group id to storage.MessageStorage
+// and sends reply message with buttons to press to get schedule.
+func handleNewMessage(
```
это можно унести в `/internal/app/entrypoint/telegram/bot.go`

goodmove commented on internal/bot/bot.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+// newBot creates bot.
+func newBot() *tgbotapi.BotAPI {
```
здесь кусок можно унести в конфиг я думаю, а кусок в main или рядом там положить

goodmove commented on internal/bot/bot.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+// menuButtonPressed checks if user pressed the button of replied message to him.
+func menuButtonPressed(callBackQuery tgbotapi.Update) bool {
```
вот эти хендлеры все я бы унес в `/internal/app/entrypoint/telegram/bot.go`

goodmove commented on internal/constants/days.go:
```diff
@@ -0,0 +1,14 @@
+package constants
+
+const (
+	Monday    string = "Monday"
```
это можно выкинуть все

goodmove commented on internal/constants/location.go:
```diff
@@ -0,0 +1,3 @@
+package constants
+
+const Location = "Asia/Novosibirsk"
```
что за location?) тогда уж `const Novosibirsk = "Asia/Novosibirsk"`

и пакет я б сделал `xtimezone` - здесь же про таймзоны речь

goodmove commented on internal/app/storage/groupStorage.go:
```diff
@@ -0,0 +1,23 @@
+package storage
+
```
```
/cmd/app/
-- config.go // классы с конфигурацией приложения
-- main.go // инициализация всех компонентов
/internal/
-- app/entrypoint/telegram/
---- handler.go // обработчик событий от бота
-- domain/schedule/
---- schedule.go // описание структур данных
---- schedule_service.go // описание операция с расписанием
---- schedule_storage.go // интерфейс для хранилища
-- infra/storage/
---- schedule_storage.go // реализация хранилища 
/pkg/
-- integration/unisite/
---- client.go // http клиент под интеграцию
---- schedule_api.go // API для интеграции и парсинга данных
-- xtime/
---- time.go // кастомные операции и константы
-- i18n/
---- ru.go // преобразование в строки на русском языке
```

goodmove replied:
так будет гораздо чище весь проект и внесение изменений будет прозрачнее

goodmove commented on internal/integration/unisite/parser.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// isHaveLessons checks if we have lessons while parsing.
+func isHaveLessons(lessons []string) bool {
```
`hasLessons`

мне кажется бессмысленная обертка

этого уже достаточно
```go
if len(lessons) > 0 {}
```

goodmove commented on internal/integration/unisite/parser.go:
```diff
@@ -0,0 +1,194 @@
+)
+
+// parseDate finds html selection with date that user chosen.
+func parseDate(date time.Time, url string) (*goquery.Selection, error) {
```
обычно сверху файла складывают публичные объявления и приватные константы, чтобы сразу можно было просмотреть интерфейс пакета

goodmove commented on internal/integration/unisite/parser.go:
```diff
@@ -0,0 +1,194 @@
+	return dateSelection, nil
+}
+
+// ParseDayLessons finds study group`s lessons for day that user chosen.
```
надо добавить уроки грамматики на present perfect :)

goodmove commented on internal/integration/unisite/api.go:
```diff
@@ -0,0 +1,20 @@
+
+// GetSchedule returns pointer to schedule of type model.Schedule
+// and custom error if http request Timeout occurred, otherwise error equals to nil.
+func GetSchedule(url, inputCallback string) (*model.Schedule, error) {
```
предлагаю оформить интерфейс по-другому:

```go
func GetSchedule(fromDate time.Time, toDate time.Time) (*model.Schedule, error)
```

вот это уже можно использовать и тестировать нормально) И нет никаких странных абстракций на уровне интеграции

goodmove commented on internal/integration/unisite/api.go:
```diff
@@ -0,0 +1,20 @@
+package unisite
```
**client.go**

```go

type HttpClient type {
  client *resty.Client
}

func NewHttpClient(config) ScheduleApi {
  ...
}

```


---

**api.go**
```go
type ScheduleApi interface {
  GetSchedule()
}

type GetScheduleResponse {}

func (httpClient *HttpClient) GetSchedule(fromDate time.Time, toDate time.Time) (*GetScheduleResponse, error) {
  ...
}
```

goodmove commented on internal/app/service/scheduleService.go:
```diff
@@ -0,0 +1,289 @@
+
+// WeekScheduleToString converts schedule of type model.Schedule to string
+// if user chosen schedule on week/next week.
+func WeekScheduleToString(
```
```
view – in/out interface
-- website
-- bot
-- ios
-- console

domain – business logic
-- service

infrastructure – 
-- storage implementation
-- integration
-- message broker


app --> *
view --> domain <-- infra
```

```go
f(studyGroupId string, schedule Schedule) => string 
```

goodmove commented on internal/app/service/scheduleService.go:
```diff
@@ -0,0 +1,289 @@
+
+// DayScheduleToString converts schedule of type model.Schedule to string
+// if user chosen schedule on today/tomorrow.
+func DayScheduleToString(
```
week = (d1,d2,d3)
day = (d1)

f: Day => String

```

func SerizalizeWeek() {
  for dayItem := range weekSchedule.items {
    f(dayItem)
  }
}
```

goodmove commented on internal/app/storage/messageStorage.go:
```diff
@@ -0,0 +1,17 @@
+	MessageData map[int64]string // includes user`s chat id and his input group id
+}
+
+func NewMessageStorage() *MessageStorage {
```
```
InMessage
- chatId
- text

OutMessage
- chatId
- text

handleMessage: (InMessage) => OutMessage

```

goodmove commented on internal/app/storage/scheduleStorage.go:
```diff
@@ -0,0 +1,56 @@
+	"time"
+)
+
+type ScheduleStorage struct {
```
```go
// view 
func handleMessage(inMessage InMessage) OutMessage {
  chatId, studyGroupId := parseFromTg(inMessage)
  from, to := parseRange()

  schedule := getSchedule(groupId, from, to)
  responseText := toView(schedule)

  return OutMessage {
    chatId: chatId,
    text: responseText,
  }
}

// service
func getSchedule(groupId, from, to) (*Schedule, error) {
  return cache(groupId, from, to) (
    getScheduleImpl(groupId, from, to)
  )
}

func getScheduleImpl(groupId, from, to) (*Schedule, error) {
  schedule, err := callApi(groupId, from, to)

 ...
  
}
```
